compiler/passes/src/remove_unreachable.rs

```diff
@@ -0,0 +1,179 @@
+            const_parameters: input
+                .const_parameters
+                .iter()
+                .map(|param| ConstParameter { type_: self.reconstruct_type(param.type_.clone()).0, ..param.clone() })
+                .collect(),
+            input: input
+                .input
+                .iter()
+                .map(|input| Input { type_: self.reconstruct_type(input.type_.clone()).0, ..input.clone() })
+                .collect(),
+            output: input
+                .output
+                .iter()
+                .map(|output| Output { type_: self.reconstruct_type(output.type_.clone()).0, ..output.clone() })
+                .collect(),
+            output_type: self.reconstruct_type(input.output_type).0,
+            block: self.reconstruct_block(input.block).0,
+            span: input.span,
+            id: input.id,
+        };
+        self.has_return = false;
+        res
+    }
+
+    fn reconstruct_constructor(&mut self, input: Constructor) -> Constructor {
+        self.has_return = false;
+        let res = Constructor {
+            annotations: input.annotations,
+            block: self.reconstruct_block(input.block).0,
+            span: input.span,
+            id: input.id,
+        };
+        self.has_return = false;
+        res
+    }
+}
+
+impl AstReconstructor for RemoveUnreachableVisitor<'_> {
+    type AdditionalInput = ();
+    type AdditionalOutput = ();
+
+    fn reconstruct_block(&mut self, input: Block) -> (Block, Self::AdditionalOutput) {
+        (
+            Block {
+                statements: input
+                    .statements
+                    .into_iter()
+                    .scan(false, |return_seen, s| {
+                        let stmt = self.reconstruct_statement(s).0;
+                        let res = (!*return_seen).then_some(stmt);
+                        *return_seen |= self.has_return;
+                        res
+                    })
+                    .filter_map(Some)
+                    .collect(),
+                span: input.span,
+                id: input.id,
+            },
+            Default::default(),
+        )
+    }
+
+    fn reconstruct_conditional(&mut self, input: ConditionalStatement) -> (Statement, Self::AdditionalOutput) {
+        let mut then_block_has_return = false;
+        let mut otherwise_block_has_return = false;
+        let previous_has_return = core::mem::replace(&mut self.has_return, then_block_has_return);
+        let then = self.reconstruct_block(input.then).0;
+        then_block_has_return = self.has_return;
+
+        let otherwise = input.otherwise.map(|otherwise| {
+            self.has_return = otherwise_block_has_return;
+            let res = Box::new(self.reconstruct_statement(*otherwise).0);
+            otherwise_block_has_return = self.has_return;
+            res
+        });
+
+        self.has_return = previous_has_return || (then_block_has_return && otherwise_block_has_return);
```

Comment:
Same here.. just a quick overview of the logic at the start of this function would go a long way.